### DISCUSSION TITLE
Combining static LINQ query with Gridify query

### DISCUSSION THREAD

**Discussion by GlodenBoy** [alirezanet/Gridify#90] [Q&A]
![image](https://user-images.githubusercontent.com/2329528/199420251-7c8b4032-5474-4fc9-a15c-80e93b407055.png)

As in the picture, I need to query x3.state == a.state,

**Comment by GlodenBoy** [comment]
![image](https://user-images.githubusercontent.com/2329528/199422252-f53a3eb7-4ec5-4e8d-a35d-7faaa10ddd20.png)

**Reply by alirezanet** [reply]
Because I don't know project architecture, I don't understand your question exactly. 
I suggest you create a small reproducible example  (including a couple of classes and a LINQ query). this way everyone can test your code to find the best solution.

**Reply by GlodenBoy** [reply]
![image](https://user-images.githubusercontent.com/2329528/199441236-c16c178c-ed0d-4917-947b-76e98512acd5.png)

Well, I don't have enough time to do a test project, the main point of this question is how to add a map, which can query the fields of the parent class

**Reply by GlodenBoy** [reply]
The usual usage is state = "xxx", but now this xxx is a field of its own parent class

**Reply by alirezanet** [reply]
query on parent class e.g:
 `AddMap("parentName", child => child.Parent.Name)` 
query => `parentName = GoldenBoy`

**Reply by GlodenBoy** [reply]
![image](https://user-images.githubusercontent.com/2329528/199442553-caeee5bd-a521-4a24-b48e-fb37a08a4951.png)
As in the example in the figure, I query a field in a list of main, and this field must be equal to a field of main.

But I have no association in the subclass, I can use the above method to query it in linq.

**Reply by alirezanet** [reply]
> The usual usage is state = "xxx", but now this xxx is a field of its own parent class

If you don't want to pass a **value** to your query, what is the point of using Gridify? your query seems static, but gridify usually gets used for dynamic queries.

**Reply by alirezanet** [reply]
In this case, I think You need a Mapper for your parent class, not your child class. this way you also can map your child's classes using `Select` in the mapper.

**Reply by GlodenBoy** [reply]
In order to better understand the example, all the writing is static, in fact, some fields are dynamic, such as userId, I use Gridify because it can use strings or custom concatenation queries, more free, with linq can not be more friendly stitching

**Reply by alirezanet** [reply]
Also, you can combine the static part of your query with gridify. 
for example, if the second line in your example is a static query you don't need any mapping for that. 

`QueryableList.ApplyFiltering( configs ).Where( // your static query) . ToList();`

**Reply by GlodenBoy** [reply]
I actually used Addmap to solve most of the problems, but now this problem is difficult for me, because I want to query the state in the subset of the parent class to determine whether it is equal to the state of the parent class, if I use the passing, then I also need to try to query the state of the current parent class, which must be a very troublesome thing, in linq I only need child.state = main.state

**Reply by alirezanet** [reply]
That's why I suggested providing a reproducible example 😊. I'm giving you a couple of different ways to solve the problem but not exactly the code you're looking for.

**Reply by GlodenBoy** [reply]
ok， I'll create a simple example, please wait

**Reply by GlodenBoy** [reply]
[ConsoleApp4.zip](https://github.com/alirezanet/Gridify/files/9918755/ConsoleApp4.zip)
Here is an example

**Comment by alirezanet** [comment]
So for having the same result as your example there is two way: 

example code :
```csharp
var list = new List<Class1>()
{
	new()
	{
		State = "ok",
		Childs = new List<Class2>()
		{
			new()
			{
				State = "ok",
				UserId = 1,
			},
			 new()
			{
				State = "error",
				UserId = 2,
			}, new()
			{
				State = "error",
				UserId = 3,
			},
			 new()
			{
				State = "error",
				UserId = 4,
			}
		}
	},
	   new()
	{
		State = "error",
		Childs = new List<Class2>()
		{
			new()
			{
				State = "ok",
				UserId = 4,
			},
			 new()
			{
				State = "error",
				UserId =1,
			}, new()
			{
				State = "error",
				UserId = 6,
			},
			 new()
			{
				State = "error",
				UserId = 1,
			}
		}
	}
}.AsQueryable();

//How to convert to Gridify
var get = list.Where(a => a.Childs.Any(x => x.UserId == 1 && x.State == a.State));

internal class Class1
{
	public string State { get; set; }

	public List<Class2> Childs { get; set; }
}

internal class Class2
{
	public string State { get; set; }
	public int UserId { get; set; }
}
```
--- 

First, if you want to always map your child State to the parent state you should not use gridify because it is a static hard-coded query:

e.g using query builder

```csharp
var qb = new QueryBuilder<Class1>()
		.AddMap("child_id", q => q.Childs.Select(c => c.UserId))
		.AddCondition("child_id = 1") // this is your dynamic part
		.Build(list);


// combine your static part
qb = qb.Where(q => q.Childs.Any(c => c.State == q.State)); 

var result = qb.ToList();
```
---

The second way, if you wanna Control the state:
```csharp
// get the state from user to control the filtering
var state = "error";

var qb = new QueryBuilder<Class1>()
		.AddMap("child_id", q => q.Childs.Select(c => c.UserId))
		.AddMap("state", q => q.State)
		.AddMap("child_state", q => q.Childs.Select(c => c.State))
		.AddCondition($"child_id = 1, child_state={state}, state={state}")  
		.Build(list);

var result = qb.ToList();

```

**Reply by GlodenBoy** [reply]
> qb = qb.Where(q => q.Childs.Any(c => c.State == q.State));

It seems that this is the most convenient behavior, and I am currently using this method, but its disadvantage is that it cannot be converted to a string for better passing.

**Reply by alirezanet** [reply]
Feature suggestions are always welcome, how do you suggest we support this behavior differently? 
(because you're not really dealing with a dynamic query so IMO it is not Gridify related)

**Reply by GlodenBoy** [reply]
My actual project has a large number of query conditions, most of these conditions are static, a few about UserId, which is based on the character Id, and there is a stitching query in the process, which is undoubtedly troublesome with linq.

I don't know much about the source code and principles of linq, so I think Gridify can extend a way to directly convert linq to string?

**Reply by alirezanet** [reply]
*I updated the last example*,
what do you mean about direct conversion? 

something like this? 

```csharp
// Map a single string to multiple different properties
var qb = new QueryBuilder<Class1>()
		.AddMap("child_id", q => q.Childs.Select(c => c.UserId))
		.AddMap("state", new[] { q => q.State,  q => q.Childs.Select(c => c.State) })
		.AddCondition($"child_id = 1, state=error")  
		.Build(list);
```

This might be possible, Although not sure if it's useful or not.

**Reply by GlodenBoy** [reply]
No, it's parsing linq queries and converting them into strings recognized by Gridify or QueryBuilder

**Reply by GlodenBoy** [reply]
The main difficulty above is that state is neither dynamic nor static, it is related to the parent, and I can't pass values unless I associate them when constructing the dataset, but this undoubtedly adds a lot of work, because this is queried from the database, they come from multiple tables, and they are related with primary keys

**Reply by alirezanet** [reply]
> No, it's parsing linq queries and converting them into strings recognized by Gridify or QueryBuilder

Well, could be an interesting feature, but honestly, I'm not sure what problem that string can solve... 
maybe with a good proposal/explanation, I add it to the next versions. 🤷‍♂️

**Reply by GlodenBoy** [reply]
For me strings are better customized than linq, I can send query conditions to the server, let the server query, and then return me the actual result, linq words currently have no particularly easy way to pass
